### PR TITLE
Added OpenCV and PCL support

### PR DESCRIPTION
**Pull request by phuicy**
I added these because they are very highly used modules for ROS.

The other changes make the readme.md a little bit more clear and comprehensive.

### CODE REVIEW

**Comment by wjwwood** [comment]
OpenCV and PCL provide their own CMake modules:
- PCL: http://pointclouds.org/documentation/tutorials/using_pcl_pcl_config.php
- OpenCV: http://docs.opencv.org/doc/tutorials/introduction/linux_gcc_cmake/linux_gcc_cmake.html#create-a-cmake-file

I would not reproduce them here, this package should be reserved for CMake modules to find things which do not provide their own CMake infrastructure and do not have CMake modules provided by CMake itself.

Can you explain, in more detail, why you need these?

File: README.md [hunk 1]
```diff
@@ -1,10 +1,29 @@
 cmake_modules
 =============
+ <img src="http://upload.wikimedia.org/wikipedia/commons/0/0a/Cmake.jpg" alt="CMake Logo" align="left" width="160"/>
```
Comment (tfoote) [inline README.md]: Images are nice but it should not be done via hotlinking. And I don't think that it's worth managing the resource locally either.

Comment (phuicy) [inline README.md]: Yeah, i entirely agree, it was more for my benefit, i couldn't remember how to resize images. I meant to remove it.

**Comment by wjwwood** [comment]
Some of the other documentation changes/improvements seem valuable, so even if we don't take the PCL and OpenCV modules, you should consider a separate pull request with some of those changes.

**Comment by phuicy** [comment]
The reason to include them here, is for ease of development. The shear number of times I have had to manually include them is infuriating. And seeing as these 2 packages are integral to the ROS community, I feel it would be worth while to include them in a package like this.

**Comment by tfoote** [comment]
Can you be a little clearer about what you need to do to include them? On my system they are in the default path.  

```
tfoote@BigFoote:/tmp$ cmake --find-package -DNAME=PCL -DLANGUAGE=C -DCOMPILER_ID=GNU -DMODE=COMPILE
-I/usr/include/vtk-5.8 -I/usr/include/pcl-1.7 -I/usr/include/eigen3 -I/usr/include/ni -I/usr/include/qhull  
tfoote@BigFoote:/tmp$ cmake --find-package -DNAME=OpenCV -DLANGUAGE=C -DCOMPILER_ID=GNU -DMODE=COMPILE
-I/usr/include/opencv  
```

**Comment by phuicy** [comment]
I have to include the find****.cmake files,  and  in the CMakLists.txt: 

``` cmake

set(CMAKE_MODULE_PATH ${CMAKE_MODULE_PATH}  ${catkin_INCLUDE_DIRS}/FindOpenCV.cmake)
find_package(OpenCV 2.4.8 REQUIRED)
include_directories(${CMAKE_SOURCE_DIR}/include
                                      ${OpenCV_INCLUDE_DIRS})
```

I have checked with 4 associates and this how they also manage the problem too. Its seems to be true for both default build locations, and user specified build/install locations. 

It might not seem much, but its harder than:

``` cmake
find_package(OpenCV 2.4.8 REQUIRED)
```

Moreover, I know 2 people who have remained in fuerte and hence rosbuild, as they couldn't figure out the fix, and wanted to avoid the issue. 

P.s. in both groovy and hydro

**Comment by phuicy** [comment]
I changed it to be less contentious, if you choose to leave it as is.
In the meantime; I will continue to use (and distribute) my version, to solve my issue.

**Comment by wjwwood** [comment]
I will review the documentation changes you now propose.

---

However, just for my own curiosity, I am concerned by this line in your above example:

``` cmake
set(CMAKE_MODULE_PATH ${CMAKE_MODULE_PATH}  ${catkin_INCLUDE_DIRS}/FindOpenCV.cmake)
```

This doesn't make any sense to me, `${catkin_INCLUDE_DIRS}` is a list of directories where header files can be found. So it will be something like this:

```
/opt/ros/hydro/include;/usr/include;/path/to/my/workspace/include
```

The above CMake will add each of those to the `CMAKE_MODULE_PATH` and append `/FindOpenCV.cmake` to the last one. This doesn't make sense because there are no CMake modules in these directories and you are not supposed to put absolute paths to `.cmake` files in the `CMAKE_MODULE_PATH`. Is this a typo? If not I would seriously challenge that this is doing what you expect it does.

**Comment by phuicy** [comment]
Therein lies the part of the problem; /path/to/my/workspace/src/package_name/include is where I have to put the find*****.cmake, which is annoying and unnecessary.  
I could just use include, but it still the same.

**Comment by wjwwood** [comment]
@phuicy that is not where cmake files belong. They shouldn't be installed unless other packages which depend on you need them. In which case they should be installed to `<prefix>/share/<package name>/cmake` or into `<prefix>/share/cmake/Modules`. They absolutely should not be installed to the `include` directories.

All of this is only matters if you need to ship a CMake module which isn't provided elsewhere, but for OpenCV and PCL there are CMake modules available when they are installed, so I don't understand why you need to do this at all?

**Comment by phuicy** [comment]
The cmakes are not on the default for myself and other I have talked to.
Thus, it makes it more portable as a singular contained module.

They could exist in a separate cmake folder in the package, but its a minor detail and mainly down to my laziness.

**Comment by wjwwood** [comment]
@phuicy That's very strange, as far as I know that shouldn't be necessary. If you can provide more details or get others to weigh in then I would consider putting them in here, but without that evidence I don't see the need.

@vrabaud any idea why @phuicy might not find OpenCV without including his own file?

**Comment by tfoote** [comment]
@phuicy What system are you running? And how were OpenCV and PCL installed? Are you using non-standard versions or a non-standard installation path? Where do the cmake files get installed?

**Comment by phuicy** [comment]
2 university pcs (different universities)(ROS and OpenCV not installed by me). 2 home PCs , 1 laptop , 1 aws server, 2 pioneers 3at, and 1 higher spec other rover.  So its not a specs issue, and the fact that its happening to multiple people, I don't think its me (although it could I normal a JAVA guy). 
Both standard and user defined installation for both ROS and openCV. Only standard for PCL.

**Comment by phuicy** [comment]
Looking at answers.ros.org, this seems to be what every one else is doing to.
However, to be fair, some people answer they way you do, but it doesn't seem to work for everyone

**Comment by tfoote** [comment]
@phuicy Could you point to specific relevant questions on answers.ros.org? And can you provide descriptions of your system so that we can try to reproduce your problem? Without enough details to reproduce your setup it's hard for us to help you get the right solution.

**Comment by vrabaud** [comment]
I remember a bug in OpenCV (2.4.8 ?) where the find*.cmake would not be installed. But that got fixed with later versions and is not present in the ROS versions of those packages.
Do you use your distro packages or the ros-hydro-opencv2 package ? What is your distro ?

**Comment by phuicy** [comment]
Its seems if it was an OpenCV issue which has been resolved now. Thanks for your help.

**Comment by vrabaud** [comment]
cool ! maybe you need to squash your commits then ? Just to clean up. Or make a new PR just for the doc.

**Comment by phuicy** [comment]
I'll make a new pull request for simplicity. 

Many thanks

**Pull request closed without merging**